src/pydicom/filebase.py

```diff
@@ -1,212 +1,302 @@
 # Copyright 2008-2020 pydicom authors. See LICENSE file for details.
 """Hold DicomFile class, which does basic I/O for a dicom file."""
 
+from collections.abc import Callable
 from io import BytesIO
-from struct import unpack, pack
+import os
+from struct import Struct
 from types import TracebackType
-from typing import BinaryIO, cast, TextIO, Any
-from collections.abc import Callable
+from typing import cast, Any, TypeVar, Protocol
+
+
+ExitException = tuple[
+    type[BaseException] | None, BaseException | None, TracebackType | None
+]
+Self = TypeVar("Self", bound="DicomIO")
+
+
+class ReadableBuffer(Protocol):
+    def read(self, size: int = ..., /) -> bytes:
+        ...  # pragma: no cover
 
-try:
-    from typing import Protocol  # added in 3.8
-except ImportError:
-    Protocol = object  # type: ignore[assignment]
+    def seek(self, offset: int, whence: int = ..., /) -> int:
+        ...  # pragma: no cover
```

Comment:
For my education, why the Ellipsis in the function arguments?  e.g. why not just `whence: int`?

Reply:
Because there's an actual default value, but for the purposes of typing what that value actually is doesn't matter.

Reply:
Also, I've found looking at the type hints in typeshed to be pretty useful and this is [pretty much taken from there](https://github.com/python/typeshed/blob/1245faf0d5f7baab32834d91606ad22ed758d7f4/stdlib/io.pyi#L70).